Bluetooth: controller: Extended scanning with LLL scheduling

andrzej-kaczmarek opened the pull request:
This is further development based on https://github.com/zephyrproject-rtos/zephyr/pull/35391.
At the moment it is capable of scanning complete ADV_EXT_IND + AUX_ADV_IND + AUX_CHAIN_IND chain with minimum Tmafs, ULL and LLL scheduling can be interleaved depending of aux offset.

The RX path for AUX is LLL is not changed too much, the only difference is that AUX scheduled from LLL is sent as separate event to ULL since it does not hold aux context, but has lll_scan instead. ULL will restore aux context for such event which was stored on previous event (i.e. on ext). This is based on an assumption that if aux was scheduled from LLL on a PDU, then next PDU that is scanned will be that scheduled PDU so there's no risk of mixing PDUs from different chains.

What does not work yet:
- not yet sure why, but if 1st aux is scheduled from last ext (via lll) in advertising event, scan does not resume after complete chain is scanned; this works fine if aux is scheduled from ull
- scanning chain in sync is not done; it works pretty much the same except different structs are passed via param so needs some more work to chain all event together

cvinayak committed:
Bluetooth: controller: Extended scanning with LLL scheduling

Extended scanning with LLL scheduling.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>
Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

cvinayak commented on subsys/bluetooth/controller/ll_sw/lll_scan.h:
```diff
@@ -31,6 +31,9 @@ struct lll_scan {
 	uint16_t duration_expire;
 	uint8_t  phy:3;
 	uint8_t  is_adv_ind:1;
+
+	/* XXX: used with lll scheduling */
+	struct lll_scan_aux *lll_aux;
```
Correct me, this is only needed as initiator to remember the CONNECT_REQ, instead move the `node_conn_rx` into `struct lll_scan` ?

andrzej-kaczmarek replied:
This is needed when aux is scheduled from LLL.
For example, if aux was scheduled from LLL on ext PDU, once that ext PDU reaches ULL we will allocate lll_scan_aux as usual. However, since we skip ULL scheduling and thus do not pass that information to LLL, we need to store it somewhere and then retrieve when scanned aux reaches ULL. Similar process is when we scheduled aux to aux from LLL, so 1st aux stores lll_scan_aux and 2nd aux retrieves it.

cvinayak reviewed (changes requested):
I was planning to develop on babblesim using `tests/bluetooth/bsim/bsim_test_advx` and modified advertising that keeps the primary PDU and auxiliary PDU at constant interval and constant offset (I have some commits in my dev branch, that I can send as draft PR for debug purposes)

Otherwise, I have some changes requested. Overall, happy with the direction this PR is going.

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:
```diff
@@ -1286,6 +1296,12 @@ static int isr_rx_scan_report(struct lll_scan *lll, uint8_t rssi_ready,
+				ftr->extra =
+					isr_rx_scan_aux_setup(lll, pdu_adv_rx);
+				if (ftr->extra) {
+					err = -EBUSY;
```
This error code does not affect caller of the function: `isr_rx_pdu`. It is used as a local flag to check if node has to be enqueued towards ULL.

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:
```diff
@@ -1232,6 +1241,7 @@ static int isr_rx_scan_report(struct lll_scan *lll, uint8_t rssi_ready,
 				uint8_t rl_idx, bool dir_report)
 {
 	struct node_rx_pdu *node_rx;
+	int err = 0;
```
Change to e.g. `bool scan_aux_en`?

ppryga replied:
```suggestion
	bool scan_aux_en = false;
```

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:
```diff
@@ -1312,8 +1328,211 @@ static int isr_rx_scan_report(struct lll_scan *lll, uint8_t rssi_ready,
+	}
 
-	return 0;
+	return err;
```
```suggestion
	return 0;
```

ppryga replied:
```suggestion
	return 0;
```

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:
```diff
@@ -1312,8 +1328,211 @@ static int isr_rx_scan_report(struct lll_scan *lll, uint8_t rssi_ready,
+		printk("ENQUEUE\n");
+		ull_rx_put(node_rx->hdr.link, node_rx);
+		ull_rx_sched();
+	} else {
```
Remove if not used.

ppryga replied:
Remove if not used.

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:
```diff
@@ -1135,6 +1138,12 @@ static inline int isr_rx_pdu(struct lll_scan *lll, struct pdu_adv *pdu_adv_rx,
 					 dir_report);
 		if (err) {
+			/* Auxiliary PDU LLL scanning has been setup */
+			if (IS_ENABLED(CONFIG_BT_CTLR_ADV_EXT) &&
```
Special handling of error seems to be not needed. There is no difference between handling of failure and success of `isr_rx_scan_report`. Also `-EBUSY` is not needed to be returned. See my comment: https://github.com/zephyrproject-rtos/zephyr/pull/36588/commits/7ce45e9ef2b7c387df09c15ac7556697660039a4#r667901999

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:
```diff
@@ -1312,8 +1328,211 @@ static int isr_rx_scan_report(struct lll_scan *lll, uint8_t rssi_ready,
+	}
+
+	/* traverse through adi, if present */
+	if (pri_hdr->adi) {
```
What if `pri_pdu` has `tgt_addr` field?

ppryga replied:
What if pri_pdu has tgt_addr field?

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:
```diff
@@ -1312,8 +1328,211 @@ static int isr_rx_scan_report(struct lll_scan *lll, uint8_t rssi_ready,
+
+	/* Determine the window size */
+	if (aux_ptr->offs_units) {
+		window_size_us = 300U;
```
Change numbers to macros. It will be easier to understand code for new person.

ppryga replied:
Change to macro please. It will make reading code easier for new people.

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:
```diff
@@ -1312,8 +1328,211 @@ static int isr_rx_scan_report(struct lll_scan *lll, uint8_t rssi_ready,
+	if (aux_ptr->offs_units) {
+		window_size_us = 300U;
+	} else {
+		window_size_us = 30U;
```
Change numbers to macros. It will be easier to understand code for new person.

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:
```diff
@@ -1286,6 +1296,12 @@ static int isr_rx_scan_report(struct lll_scan *lll, uint8_t rssi_ready,
+				ftr->extra =
+					isr_rx_scan_aux_setup(lll, pdu_adv_rx);
+				if (ftr->extra) {
+					err = -EBUSY;
```
This error code does not affect caller of the function: `isr_rx_pdu`. It is used as a local flag to check if node has to be enqueued towards ULL. 

```suggestion
					scan_aux_en = true;
```

andrzej-kaczmarek replied:
It does affect caller because it will then `return 0` which in turn will affect its called not to disable radio. This is because `isr_rx_scan_aux_setup` already disabled radio to proceed with aux scheduling, disabling radio "as usual" would overwrite that.

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:
```diff
@@ -1135,6 +1138,12 @@ static inline int isr_rx_pdu(struct lll_scan *lll, struct pdu_adv *pdu_adv_rx,
 		if (err) {
+			/* Auxiliary PDU LLL scanning has been setup */
+			if (IS_ENABLED(CONFIG_BT_CTLR_ADV_EXT) &&
+			    (err == -EBUSY)) {
```
Special handling of error seems to be not needed. There is no difference between handling of failure and success of `isr_rx_scan_report`. Also `-EBUSY` is not needed to be returned. See my comment for line 1303.

andrzej-kaczmarek replied:
see my comment for mentioned line

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:
```diff
@@ -401,6 +414,13 @@ static int isr_rx_pdu(struct lll_scan_aux *lll, uint8_t devmatch_ok,
+		/* FIXME: we need lll_aux for aux_connect_req handling, need to
+		 *        fix somehow
+		 */
+		if (!lll_aux) {
```
Correct me if I'm wrong. This will return -ENOBUFS if received ADV_EXT_IND PDU with pointer `aux_ptr` to auxiliary PDU with offset short enough to do not go through ULL (schedule next PDU from lll_scan).

andrzej-kaczmarek replied:
That is just temporary until I update it to handle aux_adv_ind scheduled from LLL on connectable, so this will change in final version.

andrzej-kaczmarek committed:
Bluetooth: controller: Fix isr_done_cleanup

We need to reset isr status before returning as otherwise irq will be
triggered over and over again.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>
Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

andrzej-kaczmarek commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -208,7 +208,8 @@
 #define BT_CTLR_MAX_CONN        CONFIG_BT_MAX_CONN
 #if defined(CONFIG_BT_CTLR_ADV_EXT) && defined(CONFIG_BT_CENTRAL)
-#define BT_CTLR_ADV_EXT_RX_CNT  1
+/* FIXME: needs more due to lll scheduling */
+#define BT_CTLR_ADV_EXT_RX_CNT  16
```
Yes, assuming ULL can dispatch scanned PDU to HCI before LLL scans another aux that was scheduled from LLL. I don't think that should be a problem though.

andrzej-kaczmarek commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:
```diff
@@ -785,3 +796,239 @@ bool isr_scan_connect_rsp_check(struct pdu_adv *pdu_tx, struct pdu_adv *pdu_rx)
+	aux_start_us = radio_end_us;
+	aux_start_us -= PKT_AC_US(pdu->len, 0, lll->phy);
+	aux_start_us += aux_offset_us;
+	aux_start_us -= lll_radio_rx_ready_delay_get(phy, 1);
```
we don't have flag for this field so I'll leave it for now as-is, don't want to change unrelated code in this PR

andrzej-kaczmarek commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:
```diff
@@ -785,3 +796,239 @@ bool isr_scan_connect_rsp_check(struct pdu_adv *pdu_tx, struct pdu_adv *pdu_rx)
+			    PKT_AC_US(pdu->len, 0, lll->phy);
+
+	radio_isr_set(isr_scan_aux_setup, node_rx);
+	radio_disable();
```
Yes, because we need different radio disable handling (we'll schedule aux scan from there) than in standard case. The return value will tell caller to skip standard radio disable routine.

andrzej-kaczmarek committed:
Bluetooth: controller: Add ull_scan_aux_is_valid_get helper

This can be used to check if pointer is a valid lll_scan_aux pointer,
e.g. if we need to decide what kind of struct is attached to RX node
for further processing.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:
```diff
@@ -556,13 +556,18 @@ static int isr_rx_pdu(struct lll_scan_aux *lll, uint8_t devmatch_ok,
 		ftr = &(node_rx->hdr.rx_ftr);
-		ftr->param = lll;
+
+		if (lll_aux) {
+			node_rx->hdr.type = NODE_RX_TYPE_EXT_AUX_REPORT;
+			ftr->param = lll_aux;
+		} else {
+			node_rx->hdr.type = NODE_RX_TYPE_EXT_AUX_REPORT_LLL;
+			ftr->param = lll_scan;
+		}
```
I'd prefer to not make too many "extra" changes here and stick to current `param` usage, it can be resolved later if really needed

andrzej-kaczmarek commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:
```diff
@@ -1287,6 +1302,19 @@ static int isr_rx_scan_report(struct lll_scan *lll, uint8_t rssi_ready,
+
+				/* FIXME figure out why ext-to-aux scheduling
+				 *       from LLL still fails sometimes...
+				 */
```
This seemed to work initially but then I found it still can fail so just commented out for now so it can be tested...

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:
```diff
@@ -777,6 +777,7 @@ static void isr_abort(void *param)
 static void isr_done_cleanup(void *param)
 {
 	if (lll_is_done(param)) {
+		radio_status_reset();
 		return;
 	}
```
Good catch, move the reset outside the return as generally the rest of the cleanup code path is missing the reset even when `lll_is_done` is false. You may send this commit as separate PR for early approval to merge.

```suggestion
{
	/* Clear radio status and events */
	lll_isr_status_reset();
	
	if (lll_is_done(param)) {
		return;
	}
```

cvinayak commented:
@andrzej-kaczmarek I have identified the issues and working on fixes, I will send commits to this PR alongwith the review requests reworked by today, please do not analyse further I send the fixes.

cvinayak committed:
Bluetooth: Controller: Fix Extended Scanning Address Resolution

Fix the received Extended Advertising PDU's address
resolution latency due to missing reset of pending IRQ in
the NVIC which caused the wait loop from not exit correctly.

Fixes #30243.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Flush Aux PDUs being in LLL scheduling

When Auxiliary PDUs are scanning by the primary scan's LLL
scheduling, then flush the Auxiliary PDUs when processing
them in the ULL instead of using the disabled callback.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Resume primary PDU scanning after Auxiliary PDU

Add implementation to resume primary PDU scanning window
after use of LLL scheduling to receive auxiliary PDUs.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
Now the scanning should be to some extent stable. I did find about a bug that in `is_abort_cb` for the `lll_scan.c`, the preemption by auxiliary scan converts the scan window into continuous scan. Will try to have a fix next week.

cvinayak committed:
Bluetooth: Controller: Fix tag name should be unique violation

Fix tag name should be unique violation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>
